docs/configuration.md
# Configuration

## Via UI

If you prefer to set up MovieMatch using a web interface, just start MovieMatch
and you will be presented with a configuration screen. The configuration will be
saved in the working directory.

## Via YAML

MovieMatch can be configured with a simple YAML document, which allows
connecting to multiple Plex servers.

Here's a simple example:

```YAML
host: 0.0.0.0
port: 8000
servers:
  - url: https://plex.foo.com
    token: abcdef12346
    useForAuth: true
  - url: https://plex.bar.com
    token: abcdef12346
    useForAuth: true
permittedAuthTypes:
  anonymous:
    - JoinRoom
  plex:
    - JoinRoom
  plexFriends:
    - JoinRoom
    - CreateRoom
  plexOwner:
    - JoinRoom
    - CreateRoom
    - DeleteRoom
    - ResetRoom
    - Admin
```

MovieMatch will read the config from `config.yaml` by default.


RELEASE_NOTES.md
# Release Notes

## v2.0.0-beta.2

- [Bugfix] You can now toggle the user menu by tapping the avatar multiple times
- [Bugfix] Like / Dislike buttons now disappear when there are no more cards
- [Bugfix] error messages for Login / Join / CreateRoom are now shown to the
  user
- [Bugfix] Boolean filters now work as expected
- [Bugfix] `requirePlexTvLogin` config option now works as expected

## v2.0.0-beta.1

- [CI]: Run end-to-end tests everywhere
- [Docs]: Add information about common crashes to the FAQ
- [Testing] Improvements to end-to-end tests
- [Testing] Include the UI in linting tests (tsc)
- [Compat] Rename `basePath` to `rootPath`
- [Bugfix] The server is now notified when a user logs out and will remove them
  from the room
- [Bugfix] More robust messages for filter values and setup
- [UI] Refactor the store from `useReducer` + `Context` to Redux
- [UI] Refactor to CSS Modules
- [Bugfix] Fix #74
- [Feature] Implement a new User menu showing other users in a room, as well as
  user progress
- [Feature] Use icons from [Iconic](https://iconic.app)
- [Feature] Show matches as a grid on larger screens
- [Feature] Show a like / dislike button on larger screens
- [Feature] Add an 'open in plex' link to the card stack cards

## v2.0.0-alpha.7

- [Bugfix] Improve Card Stack for different screen sizes
- [Bugfix] MovieMatch would fail to start on Windows
- [Bugfix] Having Reduced Motion enabled would break the Card Stack
- [Bugfix] The server is no longer sent duplicate ratings when a card is
  dismissed
- [Feature] Provide ARM64 Linux binary
- [CI] Run end-to-end tests for macOS, Linux, and Windows against the release
  binary

## v2.0.0-alpha.6

- Fix and issue with the release artefact for alpha 5

## v2.0.0-alpha.5

- [Bugfix]: min-height regression in iOS
- [Bugfix]: Movie card text was duplicated
- [Bugfix] Ensure all references work with `x-forwarded-prefix` and `ROOT_PATH`
- [Bugfix] Don't show the card description twice
- [CI] Compile the bundle before creating a release
- [Bugfix] Don't delay rendering of the Create Room filters `+` button
- [Bugfix] Don't allow an item to be rated twice by the same user.
- [Housekeeping] Remove built-in dev server (using Snowpack now)
- [Testing] Validate plexApi.getLibraryItems()
- [Config] Rename `BASE_PATH` back to `ROOT_PATH` (it was causing too much
  confusion)
- [Feat] Don't overscroll the page (`overscroll-behavior` isn't supported in
  Safari yet)
- [Feat] Scroll to the beginning of the matches list when the sort type is
  changed
- [Bugfix] The server is notified when a user leaves a room deliberately
- [CI] Now using [Just](https://github.com/casey/just) for scripting

## v2.0.0-alpha.4

- Upgrade Deno to 1.8.2
- Don't fail entire test suite if TEST_PLEX_URL and TEST_PLEX_TOKEN are missing
  (fixes failing tests on PRs)
- Move to [Snowpack](https://www.snowpack.dev) for UI tooling
- Improvements to `<CardStack />` component (#58)
- Improvements to `<Card />` component (#9)

## v2.0.0-alpha.3

- Publish the JS bundle to the release artefacts (Fixes an issue with Docker
  images not building for ARM64 under QEMU)
- Ask for permission before using read/write/env/net if run with `deno run`
  without any permission flags
- Fix a regression in room creation filters
- Fixes to the CI pipeline

## v2.0.0-alpha.2

- Rewrite configuration handling with extensive testing and stability
  improvements
- Default log level to `DEBUG`
- Implement AutoSuggest input for filter values
- Disable focus ring until a user presses the tab key (that'd be the first thing
  a keyboard user would press), and disable focus ring again when a pointer is
  used
- Implement match sorting based on "most recent" or "most likes" (#20)
- Upgrade to Deno v1.8.0

## v2.0.0-alpha.1

### New features

- Provide standalone binaries for Linux, macOS, and Windows. (No need to install
  Docker or Deno)
- Support configuring MovieMatch in the application, providing a more
  user-friendly setup experience.
- Allow configuration via environment variables and a more structured (and safe)
  YAML format for more advanced use cases
- Implement a new User Interface
- Support sharing the room URL more easily (#7)
- Support for logging out or leaving a room by tapping on your user name
- Show media details in the card (#9)
- Allow users to sign on with Plex (#14)
- Add an option that restricts login only to users signed in with Plex
- Allow filtering when creating rooms (#21)

### Bug fixes

- The UI is now compiled with [`esbuild`](https://esbuild.github.io), which
  provides improved support for older browsers. (#28)
- Improved logging throughout
- The logger will automatically redact any Plex URL or Token, making it safer
  for sharing logs when reporting issues

### Known issues

- The UI for filtering is unstyled
- The UI for configuration is quite ugly
- Card details cannot be scrolled
- Card details does not include useful metadata like rating, genre, actors, etc.
- Swiping the card stack can sometimes be laggy or get stuck
- When a third user matches an item the item is duplicated for all users
- The user needs to join the room each time they launch the app
- iOS tries to autofill Room Name as a regular name (not a problem on Android)
- A user leaving a room is not correctly detected
- UI is broken on Edge for Android

## v1.10.0

- Update Deno to v1.7.5
- Don't crash if the `PLEX_URL` is passed with a trailing `/` (Fixes #52)

## v1.9.0

- Add French translations (Thanks @EVOTk)

## v1.8.0

- Add collection filter (Thanks @BobWithHair!)

## v1.7.0

- Update nl translations (Thanks @jasperstrijp)
- Publish ARM64 Docker image (#27)

## v1.6.0

- [Bugfix] Environment variables exported to deno were not trimmed of
  whitespace, whereas values specified in an `.env` file were. This release
  ensures environment variables passed using both methods are trimmed. (#36)

## v1.5.0

- Fix for #31
- Possible fix for #33
- Debugging for #26

## v1.4.0

- Fix for #29

## v1.3.0


internal/app/moviematch/config.ts
import { parse, stringify } from "encoding/yaml.ts";
import * as log from "log/mod.ts";
import { assert } from "util/assert.ts";
import { join } from "path/posix.ts";
import { Config, LibaryTypes, LibraryType } from "/types/moviematch.ts";
import { addRedaction } from "/internal/app/moviematch/logger.ts";
import { isRecord } from "/internal/app/moviematch/util/is_record.ts";

export function verifyConfig(
  value: unknown,
  strict?: boolean,
): asserts value is Config {
  isRecord(value, "config");

  if (value.hostname) {
    assert(typeof value.hostname === "string", "hostname must be a string");
  } else {
    value.hostname = "0.0.0.0";
  }

  if (value.port) {
    assert(typeof value.port === "number", "port must be a number");
  } else {
    value.port = 8000;
  }

  if (value.logLevel) {
    assert(
      typeof value.logLevel === "string" &&
        Object.keys(log.LogLevels).includes(value.logLevel),
      `logLevel must be one of these: ${Object.keys(log.LogLevels).join(", ")}`,
    );
  } else {
    value.logLevel = "INFO";
  }

  assert(
    Array.isArray(value.servers),
    `servers must be an Array`,
  );

  if (strict) {
    assert(value.servers.length > 0, `At least one server must be configured`);
  }

  for (const server of value.servers) {
    isRecord(server, "server");

    if (server.type) {
      assert(
        server.type === "plex",
        `"plex" is the only valid server type. Got "${server.type}"`,
      );
    } else {
      server.type = "plex";
    }

    assert(typeof server.url === "string", `a server url must be specified`);

    addRedaction(server.url);

    assert(
      typeof server.token === "string",
      `a server token must be specified`,
    );

    addRedaction(server.token);

    if (server.libraryTitleFilter) {
      assert(
        Array.isArray(server.libraryTitleFilter) ||
          typeof server.libraryTitleFilter === "string",
        `libraryTitleFilter must be a list of strings or a string.`,
      );

      if (typeof server.libraryTitleFilter === "string") {
        server.libraryTitleFilter = [server.libraryTitleFilter];
      }
    }

    if (server.libraryTypeFilter) {
      assert(
        Array.isArray(server.libraryTypeFilter) ||
          typeof server.libraryTypeFilter === "string",
        `libraryTypeFilter must be a list of strings or a string.`,
      );

      if (typeof server.libraryTypeFilter === "string") {
        server.libraryTypeFilter = [server.libraryTypeFilter];
      }

      for (const libraryType of (server.libraryTypeFilter as string[])) {
        assert(
          LibaryTypes.includes(libraryType as LibraryType),
          `libraryTypeFilter(s) must be one of ${
            LibaryTypes.join(", ")
          }. Instead you entered "${libraryType}"`,
        );
      }
    }

    if (server.linkType) {
      const validLinkTypes = ["app", "webLocal", "webExternal"];
      assert(
        typeof server.linkType === "string" &&
          validLinkTypes.includes(server.linkType),
        `linkType must be one of these: ${
          validLinkTypes.join(", ")
        }. Instead, it was "${server.linkType}"`,
      );
    }
  }

  if (value.basePath) {
    assert(
      typeof value.basePath === "string",
      "basePath must be a string",
    );
    assert(value.basePath !== "/", 'basePath must not be "/"');
  } else {
    value.basePath = "";
  }

  if (value.basicAuth) {
    isRecord(value.basicAuth, "basicAuth");
    assert(
      typeof value.basicAuth.userName === "string",
      "basicAuth.userName must be a string",
    );
    assert(
      typeof value.basicAuth.password === "string",
      "basicAuth.password must be a string",
    );
  }

  if (value.requirePlexTvLogin) {
    assert(
      typeof value.requirePlexTvLogin === "boolean",
      'requirePlexTvLogin must be "true" or "false"',
    );
  } else {
    value.requirePlexTvLogin = false;
  }

  if (value.tlsConfig) {
    isRecord(value.tlsConfig);
    assert(typeof value.tlsConfig.certFile === "string");
    assert(typeof value.tlsConfig.keyFile === "string");
  }
}

export class ConfigFileNotFound extends Error {}
export class InvalidConfigurationError extends Error {}
export class ConfigReloadError extends Error {}

const canReadEnv = Deno.permissions &&
  (await Deno.permissions.query({ name: "env" })).state === "granted";

const getTrimmedEnv = (key: string) => {
  const value = Deno.env.get(key);
  if (value) return value.trim();
};

function readConfigFromEnv() {
  const HOST = getTrimmedEnv("HOST");
  const PORT = getTrimmedEnv("PORT");
  const LOG_LEVEL = getTrimmedEnv("LOG_LEVEL");
  const BASE_PATH = getTrimmedEnv("BASE_PATH");
  const PLEX_URL = getTrimmedEnv("PLEX_URL");
  const PLEX_TOKEN = getTrimmedEnv("PLEX_TOKEN");
  const AUTH_USER = getTrimmedEnv("AUTH_USER");
  const AUTH_PASS = getTrimmedEnv("AUTH_PASS");
  const REQUIRE_PLEX_LOGIN = getTrimmedEnv("REQUIRE_PLEX_LOGIN");
  const LIBRARY_TYPE_FILTER = getTrimmedEnv("LIBRARY_TYPE_FILTER");
  const LIBRARY_TITLE_FILTER = getTrimmedEnv("LIBRARY_TITLE_FILTER");
  const TLS_CERT = getTrimmedEnv("TLS_CERT");
  const TLS_KEY = getTrimmedEnv("TLS_KEY");
  const MOVIE_LINK_TYPE = getTrimmedEnv("MOVIE_LINK_TYPE");

  const envConfig: Partial<Config> = {
    hostname: HOST,
    port: PORT ? Number(PORT) : undefined,
    logLevel: LOG_LEVEL as Config["logLevel"],
    basePath: BASE_PATH,
    requirePlexTvLogin: REQUIRE_PLEX_LOGIN === "1",
    ...(AUTH_USER && AUTH_PASS
      ? {
        basicAuth: {
          userName: AUTH_USER,
          password: AUTH_PASS,
        },
      }
      : {}),
    ...(TLS_KEY && TLS_CERT
      ? {
        tlsConfig: {
          certFile: TLS_CERT,
          keyFile: TLS_KEY,
        },
      }
      : {}),
    ...(PLEX_URL && PLEX_TOKEN
      ? {
        servers: [{
          type: "plex",
          url: PLEX_URL.replace(/\/$/, ""), // Trim trailing '/' character (#52)
          token: PLEX_TOKEN,
          ...(LIBRARY_TITLE_FILTER
            ? {
              libraryTitleFilter: LIBRARY_TITLE_FILTER.split(","),
            }
            : {}),
          libraryTypeFilter: (LIBRARY_TYPE_FILTER?.split(",") ??
            ["movie"]) as LibraryType[],
          linkType: MOVIE_LINK_TYPE as Config["servers"][number]["linkType"],
        }],
      }
      : {}),
  };

  return envConfig;
}

let configPath: string;
let cachedConfig: Config;

export function getConfig() {
  if (!cachedConfig) {
    throw new Error(`getConfig was called before the config was loaded.`);
  }
  return cachedConfig;
}

export async function loadConfig(path?: string) {
  let config: Partial<Config> = {};

  try {
    const yamlConfigPath = path ?? join(Deno.cwd(), "config.yaml");
    const yamlConfigRaw = await Deno.readTextFile(yamlConfigPath);

    const yamlConfig = parse(yamlConfigRaw);
    isRecord(yamlConfig, yamlConfigPath);

    config = yamlConfig;
    configPath = yamlConfigPath;
  } catch (err) {
    if (path) {
      throw new ConfigFileNotFound(`${path} does not exist`);
    }

    log.info(
      "No default config file found. Starting up with defaults until configured.",
    );

    config = {
      port: 8000,
      hostname: "0.0.0.0",
      servers: [],
      logLevel: "DEBUG",
    };
  }

  try {
    const envConfig = canReadEnv ? readConfigFromEnv() : {};

    config = { ...config, ...envConfig };

    verifyConfig(config);

    for (const server of config.servers) {
      addRedaction(server.token);
      addRedaction(server.url);
    }

    cachedConfig = config;

    return config;
  } catch (err) {
    throw new InvalidConfigurationError(err.message);
  }
}

export async function updateConfiguration(config: Record<string, unknown>) {
  cachedConfig = config as unknown as Config;
  const yamlConfig = stringify(config, { indent: 2 });
  await Deno.writeTextFile(
    configPath ?? join(Deno.cwd(), "config.yaml"),
    yamlConfig,
  );
}
